File: go-backend-interview-prep/07_io_files_networking/06_json_encode_decode/main.go
package main

import (
	"encoding/json"
	"fmt"
)

// User demonstrates struct tags.
type User struct {
	Name    string `json:"name"`
	Age     int    `json:"age,omitempty"`
	Email   string `json:"email"`
	IsAdmin bool   `json:"-"` // never included in JSON
}

// Address has nested struct.
type Profile struct {
	User    User   `json:"user"`
	City    string `json:"city"`
	Country string `json:"country,omitempty"`
}

func main() {
	// --- Example 1: Marshal struct to JSON ---
	fmt.Println("=== Marshal ===")
	u := User{Name: "alice", Age: 30, Email: "alice@example.com", IsAdmin: true}
	data, _ := json.Marshal(u)
	fmt.Println("  json:", string(data))

	// --- Example 2: Pretty print ---
	fmt.Println("\n=== Pretty print ===")
	pretty, _ := json.MarshalIndent(u, "  ", "  ")
	fmt.Println(string(pretty))

	// --- Example 3: Unmarshal JSON to struct ---
	fmt.Println("\n=== Unmarshal ===")
	jsonStr := `{"name":"bob","age":25,"email":"bob@example.com"}`
	var u2 User
	json.Unmarshal([]byte(jsonStr), &u2)
	fmt.Printf("  parsed: %+v\n", u2)
	fmt.Println("  IsAdmin (skipped):", u2.IsAdmin) // always false

	// --- Example 4: omitempty ---
	fmt.Println("\n=== omitempty ===")
	empty := User{Name: "charlie", Age: 0, Email: "c@x.com"}
	data, _ = json.Marshal(empty)
	fmt.Println("  age=0 omitted:", string(data))

	// --- Example 5: nested struct ---
	fmt.Println("\n=== Nested struct ===")
	p := Profile{
		User:    User{Name: "dana", Age: 28, Email: "dana@x.com"},
		City:    "NYC",
		Country: "",
	}
	data, _ = json.Marshal(p)
	fmt.Println("  json:", string(data))

	// --- Example 6: dynamic JSON with map ---
	fmt.Println("\n=== Dynamic JSON (map) ===")
	dynamic := map[string]any{
		"status": "ok",
		"count":  42,
		"tags":   []string{"go", "json"},
	}
	data, _ = json.Marshal(dynamic)
	fmt.Println("  json:", string(data))

	// Unmarshal into map
	var result map[string]any
	json.Unmarshal(data, &result)
	fmt.Printf("  parsed map: %v\n", result)
}

File: go-backend-interview-prep/07_io_files_networking/06_json_encode_decode/main.py
"""JSON encode / decode -- Python equivalent of the Go example."""

import json
from dataclasses import dataclass, asdict, field


@dataclass
class User:
    name: str
    age: int = 0
    email: str = ""
    # is_admin is excluded from JSON (like json:"-")
    is_admin: bool = field(default=False, repr=False)

    def to_json_dict(self) -> dict:
        d = asdict(self)
        del d["is_admin"]       # skip like json:"-"
        if d["age"] == 0:
            del d["age"]        # omitempty
        return d


@dataclass
class Profile:
    user: User
    city: str = ""
    country: str = ""

    def to_json_dict(self) -> dict:
        d = {"user": self.user.to_json_dict(), "city": self.city}
        if self.country:
            d["country"] = self.country
        return d


def main() -> None:
    # --- Example 1: Marshal (dumps) ---
    print("=== Marshal ===")
    u = User(name="alice", age=30, email="alice@example.com", is_admin=True)
    data = json.dumps(u.to_json_dict())
    print(f"  json: {data}")

    # --- Example 2: Pretty print ---
    print("\n=== Pretty print ===")
    pretty = json.dumps(u.to_json_dict(), indent=2)
    print(pretty)

    # --- Example 3: Unmarshal (loads) ---
    print("\n=== Unmarshal ===")
    json_str = '{"name":"bob","age":25,"email":"bob@example.com"}'
    parsed = json.loads(json_str)
    u2 = User(**{k: parsed.get(k, d) for k, d in [("name", ""), ("age", 0), ("email", "")]})
    print(f"  parsed: {u2}")
    print(f"  is_admin (skipped): {u2.is_admin}")

    # --- Example 4: omitempty ---
    print("\n=== omitempty ===")
    empty = User(name="charlie", age=0, email="c@x.com")
    data = json.dumps(empty.to_json_dict())
    print(f"  age=0 omitted: {data}")

    # --- Example 5: nested struct ---
    print("\n=== Nested struct ===")
    p = Profile(user=User(name="dana", age=28, email="dana@x.com"), city="NYC")
    data = json.dumps(p.to_json_dict())
    print(f"  json: {data}")

    # --- Example 6: dynamic JSON (dict) ---
    print("\n=== Dynamic JSON (dict) ===")
    dynamic = {"status": "ok", "count": 42, "tags": ["go", "json"]}
    data = json.dumps(dynamic)
    print(f"  json: {data}")

    result = json.loads(data)
    print(f"  parsed dict: {result}")


if __name__ == "__main__":
    main()
